invokeai/frontend/web/src/features/gallery/components/ImageGrid/GalleryImageGrid.tsx

```diff
@@ -66,97 +63,90 @@ const GalleryImageGrid = () => {
     },
   });
 
-  const [didInitialFetch, setDidInitialFetch] = useState(false);
-
   const dispatch = useAppDispatch();
 
-  const {
-    galleryImageMinimumWidth,
-    imageNames: imageNamesAll, //all images names loaded on main tab,
-    total: totalAll,
-    selectedBoardId,
-    galleryView,
-    isLoading: isLoadingAll,
-  } = useAppSelector(selector);
-
-  const { data: imagesForBoard, isLoading: isLoadingImagesForBoard } =
-    useListBoardImagesQuery(
-      { board_id: selectedBoardId },
-      { skip: selectedBoardId === 'all' }
-    );
+  const { selectedBoardId, galleryView } = useAppSelector(selector);
+
+  useEffect(() => {
+    setImageList([]);
+    setOffset(0);
+    setLimit(INITIAL_IMAGE_LIMIT);
+  }, [selectedBoardId]);
+
+  const { data: imageListResponse, isLoading: isLoading } = useListImagesQuery({
+    categories: galleryView === 'images' ? IMAGE_CATEGORIES : ASSETS_CATEGORIES,
+    is_intermediate: false,
+    offset: 0,
+    limit: INITIAL_IMAGE_LIMIT,
+    ...(selectedBoardId === 'all' ? {} : { board_id: selectedBoardId }),
+  });
 
-  const imageNames = useMemo(() => {
-    if (selectedBoardId === 'all') {
-      return imageNamesAll; // already sorted by images/uploads in gallery selector
-    } else {
-      const categories =
-        galleryView === 'images' ? IMAGE_CATEGORIES : ASSETS_CATEGORIES;
-      const imageList = (imagesForBoard?.items || []).filter((img) =>
-        categories.includes(img.image_category)
+  const { data: paginatedData } = useListImagesQuery(
+    {
+      categories:
+        galleryView === 'images' ? IMAGE_CATEGORIES : ASSETS_CATEGORIES,
+      is_intermediate: false,
+      offset,
+      limit,
+      ...(selectedBoardId === 'all' ? {} : { board_id: selectedBoardId }),
+    },
+    { skip: offset === 0 }
+  );
+
+  useEffect(() => {
+    if (imageListResponse) setImageList(imageListResponse.items);
+  }, [imageListResponse]);
+
+  useEffect(() => {
+    if (paginatedData) {
+      dispatch(
+        imagesApi.util.updateQueryData(
+          'listImages',
+          {
+            categories:
+              galleryView === 'images' ? IMAGE_CATEGORIES : ASSETS_CATEGORIES,
+            is_intermediate: false,
+            offset: 0,
+            limit: INITIAL_IMAGE_LIMIT,
+            ...(selectedBoardId === 'all' ? {} : { board_id: selectedBoardId }),
+          },
```

Comment:
This is basically telling the RTK query cache to put the paginated list at the end of whatever is cached for this endpoint call (which matches the original fetch). Brittle....but works.

Reply:
tl;dr: at bottom

This is effectively recreating the behavior described here: https://redux-toolkit.js.org/rtk-query/api/createApi#merge

In the example, `serializeQueryArgs()` is used to create a single, static cache key and cache. That function could also generate a more dynamic key, like `${board_id}.${galleryView}`, which would create separate caches for each of of the possible combinations of board & view.

While your implementation and the one from the example do help to unify the endpoints, the core problem - keeping each cache updated as we move images to and from boards - still exists. 

Say you move an image from one board to another. We need to do a number of things:

1. Update the "from" board and "to" boards' caches (like, the actual cache of boards, not image lists)

RTK Query cache invalidation is fine for this. 

We just invalidate those two boards' cache tags and it auto-refetches. It's trivial to do optimistic updates on the image count, though cover image is harder to do optimistically. But just fetching those two boards' DTOs is cheap.

2. Remove the image from the "from" board's image list

We can't use cache invalidation here be cause doing we may be X images down in the list and invalidating the cache would require re-fetching all images currently cached.

So we need to use `updateQueryData` to surgically remove the image from the appropriate image list cache, and skip actual re-fetching.

3. Add the image to the "to" board's image list

This is the tricky one. We cannot use cache invalidation for the same reason as # 2 - we need to insert the image into the appropriate cache manually. 

Ok, so we `updateQueryData` to push it into the cache, but there's an issue - we don't know if the current "page" we are on should include this image. If we did just append it, and it wasn't supposed to be visible, then not only are we showing images out of order, but when you load the next page of images for that view, you'll miss an image (because the offset is 1 more than it should be due to the appended image).

We need to sort that cache, then check if the image's timestamp is within the time range represented by the cached images, and only insert the image if its timestamp is within that range. If it is outside the range, we don't need to append it.

---

Similar challenges arise when we consider the "No Board" board.

With multiple-image-operations, it gets a lot harder, because we need to do all of the above checks for every image that changed.

It's possible to use the `createEntityAdapter` to wrap the RTK Query cache, and that does help (it handles sorting and provides those helpful entity CRUD methods), but there's still a lot of manual updating needed.

I've tried this approach and ended up finding it really difficult to reason about what was happening.

Other unsuccessful approaches:
- Hybrid approach using thunks & `createEntityAdapter` to fetch & normalize the images, but also put every DTO received into the RTK Query cache for `getImageDTO`. That gives our `useGetImageDTO()` to access to the data fetched with the thunks

- Minimal manually updated cache - an array of image names, keyed to the view, inside the gallery slice. Each cache also held the network request status. This was easier to keep up to date with a few simple redux actions, but because it was just the image names, it wasn't possible to do the requisite sorting. Once you handle that, it's no better than the other strategies. Just crappily reinventing the wheel 😅 

---

So all that said (you've been to a lot of my TED talks, thanks for the support Mary!), I think doing this is feasible with two approaches:

1. Manually updating all the caches for all of the different image operations. Pretty complicated and very foot-gun-able, but very efficient. Hate it, way too brittle. ❌ 

2. Don't worry about the boards we are not on. This means the only manual cache updates we ever do are the simple filter-image-from-list updates. If we take an action that would invalidate the cache on another board, so be it - we just re-fetch it when the user clicks it again, and they have to scroll a bit. 

    I still hate it, because I know when we invalidate one of the caches, the invalidated data is probably like very, very close to correct. But this seems like the happy medium. ✅

Reply:
Oh, I forgot, even if we ignore unselected boards, there's also the case where the user drags an image from somewhere other than the gallery onto the selected board... so we still need to handle merging data into a given board. blerggh

Reply:
Ok, so I was going to work from this PR but it was easier to just start anew. I've implemented the same caching behaviour you had here with a few edge cases handled: https://github.com/invoke-ai/InvokeAI/pull/3791

Reply:
Ok that sounds good, I'll close this one out